pkgs/applications/window-managers/phosh/default.nix

```diff
@@ -0,0 +1,150 @@
+{ stdenv
+, pkgs
+, fetchFromGitLab
+, substituteAll
+, meson
+, ninja
+, pkgconfig
+, wrapGAppsHook
+, libhandy
+, pulseaudio
+, glib
+, gtk3
+, gnome3
+, gcr
+, pam
+, upower
+, wayland
+, dbus
+, xvfb_run
+, phoc
+, feedbackd
+, squeekboard
+, networkmanager
+, polkit
+, elogind
+, libsecret
+, git
+, writeText
+, makeWrapper
+, makeDesktopItem
+}:
+
+let
+  gvc = fetchFromGitLab {
+    domain = "gitlab.gnome.org";
+    owner = "GNOME";
+    repo = "libgnome-volume-control";
+    rev = "ae1a34aafce7026b8c0f65a43c9192d756fe1057";
+    sha256 = "0a4qh5pgyjki904qf7qmvqz2ksxb0p8xhgl2aixfbhixn0pw6saw";
+  };
+
+  # The upstream desktop file in phosh repo points to a stub executable
```

Comment:
It would also be nice to list why do we need this and upstream apparently think it is unnecessary. Do they launch it some different way?

Reply:
Looking at the commit message where this file was added: https://source.puri.sm/Librem5/phosh/-/commit/c3ed3787b5186bb6b78db4c55bfda2f49e1a724a

Reply:
Hmm, should not we only need it if we did not have Squeakboard? See the last point in https://source.puri.sm/Librem5/phosh/-/merge_requests/369

Reply:
Reading the merge request description, we shouldn't install the stub, or can just ignore it? (backref https://github.com/NixOS/nixpkgs/pull/88767#discussion_r429652505)

Reply:
Ok, so I'll ignore the OSK file and try to use https://github.com/NixOS/nixpkgs/pull/88767#discussion_r429652683 . It's ok?

Reply:
After patching `$out/usr/local/share/applications/sm.puri.Squeekboard.desktop` to point to the absolute path of squeekboard,, it seems that Phosh doesn't find the keyboard. I don't know if the cause is that squeekboard.desktop doesn't have:

```
 X-GNOME-Autostart-Phase=Panel
 X-GNOME-Provides=inputmethod
 X-GNOME-Autostart-Notify=true
 X-GNOME-AutoRestart=true
```

Reply:
It seems phosh controls squeekboard visibility via dbus: https://source.puri.sm/Librem5/phosh/-/blob/master/src/osk-manager.c#L17. The `.desktop` file should be unrelated.

One should be able to show/hide the keyboard with the following commands:
```
$ busctl call --user sm.puri.OSK0 /sm/puri/OSK0 sm.puri.OSK0 SetVisible b true
$ busctl call --user sm.puri.OSK0 /sm/puri/OSK0 sm.puri.OSK0 SetVisible b false
```

Reply:
I've searched a bit more in the squeekboard repo and I think the problem is the `data` folder isn't copied to the output. It contains this [dbus file](https://source.puri.sm/Librem5/squeekboard/-/blob/master/data/dbus/sm.puri.OSK0.xml) that matches 100% with your comment.

In order to have the files copied I think we need to add https://source.puri.sm/Librem5/squeekboard/-/blob/master/.gitlab-ci.yml#L38 . I've tried `mesonFlags = [ "-Ddepdatadir=$(out)/usr/share" ];` without success :(

Reply:
Meson does not interpolate its flags, you would need to use Nix's `${placeholder "out"}` instead.

Reply:
`depdatadir` defaults to `datadir`, which we set, though:

https://source.puri.sm/Librem5/squeekboard/-/blob/81961172691514c8a72388c538e111dcea327d1e/meson.build#L47-52

The issue is likely that the dbusdir is not used anywhere:

https://source.puri.sm/search?utf8=%E2%9C%93&search=dbusdir&group_id=&project_id=626&search_code=true&repository_ref=81961172691514c8a72388c538e111dcea327d1e&nav_source=navbar

Reply:
Ok. Now I see how this is supposed to work: looking at the [Libre 5 base image](https://source.puri.sm/Librem5/librem5-base/-/tree/master/default) you can see they add [sm.puri.OSK0.desktop](https://source.puri.sm/Librem5/librem5-base/-/blob/master/default/sm.puri.OSK0.desktop) that executes [osk-wayland](https://source.puri.sm/Librem5/librem5-base/-/blob/master/default/osk-wayland).

So, I think we need to decide if we want to install squeekboard + the desktop file with Phosh (like the package from alpine https://git.alpinelinux.org/aports/tree/community/phosh) or let the user decide to configure this files manually in their `configuration.nix`

Reply:
I think tracking the upstream base is better choice, we can have phosh nixos module.

Reply:
I checked a bit how to write a nixos module and I'm thinking to:

1. Add back sm.puri.OSK0.desktop to phosh derivation
2. In the nixos module install phosh, squeekboard and `environment.pathsToLink` to `/share/applications`

Am I missing something (probably yes...)? :P

Reply:
Not sure if phosh is a good place for the keyboard module (other than the stub that is there by default). I would expect either a `librem5-base` package, or ad-hoc packages using `makeDesktopItem`.

I think `pathsToLink` are already handled by one of the xdg modules.

Reply:
I created the package `phosh-ui` that add some configuration to make it work. I discarded using the `librem5-base` because it contains some configurations that only needed for their hardware.